dbcan modules

jmattock5 opened the pull request:
Modules to download the dbCAN database and to run dbCAN on a protein input based off of those in mettannotator. 

Currently formatted to run on the pre-existing version of the dbCAN database on the ftp. Once that's been updated lines 24-26 in `modules/ebi-metagenomics/dbcan/dbcandb/main.nf` need deleting (I formatted the new db to have the correct name before compressing so it no longer needs renaming).

I've included a mini dbCAN database for the nf-test for the dbCAN module

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcandb/main.nf:
```diff
@@ -0,0 +1,48 @@
+process DBCANDB {
+    tag "dbCAN $params.db_version"
```
tag with meta.id, as the rest of the nf-core modules

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcandb/main.nf:
```diff
@@ -0,0 +1,48 @@
+script:
+    def args = task.ext.args ?: ''
+    """
+    wget "${params.db_ftp_link}"
```
as we talked, define the db_version and/or url link in the def section above and use that groovy parameter in here instead

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcandb/meta.yml:
```diff
@@ -0,0 +1,26 @@
+      documentation: "https://www.gnu.org/software/wget/manual/"
+      tool_dev_url: "None"
+      licence: ["GPL v3"]
+output:
```
add inputs section..mostly copy paste from another nf-core module (dont forget the meta!)

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/meta.yml:
```diff
@@ -0,0 +1,46 @@
+  - faa:
+      type: file
+      description: Protein FASTA file containing query sequences
+      pattern: "*.{fasta,faa,fa}"
```
can it also work with .gz fasta files? if so add those extra extensions here

jmattock5 replied:
Checked and it does not

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/main.nf:
```diff
@@ -0,0 +1,51 @@
+
+    stub:
+    def args = task.ext.args ?: ''
+    def VERSION = "${params.dbcan_version}"
```
make sure there is no other way to get the dbcan version through the command line, before you hard code it here

jmattock5 replied:
I did, annoyingly there is a version option but it just prints nothing

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/main.nf:
```diff
@@ -0,0 +1,51 @@
+process DBCAN {
+    label 'process_medium'
```
tag with meta.id

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/tests/main.nf.test:
```diff
@@ -0,0 +1,45 @@
+    tag "modules"
+    tag "modules_nfcore"
+    tag "dbcan"
+
```
you are probably missing the setup step, where you run the dbcandb module first to create the db that dbcan then uses. see here for an example:t
https://github.com/nf-core/modules/blob/master/modules/nf-core/diamond/blastp/tests/main.nf.test
diamond/blastp first requires diamond/createdb to be run

jmattock5 replied:
As I said in another comment, the dbcandb module downloads the reference database, no other setup needed 😄

mberacochea replied:
Yeah, the DB doesn't have to be built... it already is.

vagkaratzas reviewed (changes requested):
Some starting comments here and there. dbcandb will also probably need its own little nf-test. Except if the database it downloads is huge I guess. Maybe if there is a smaller version?

jmattock5 commented:
I think the database is too big to download as a test but if @mberacochea agrees we could also upload the test db to the ftp and I could make a test with that.

Think I've addressed all of your comments Vangelis. Just seen that the nf-test has failed - made a new output file that it's never made before even though I've run it countless times 🤦‍♀️ will sort that next week.

vagkaratzas reviewed (changes requested):
Some extra comments

mberacochea commented:
> I think the database is too big to download as a test but if @mberacochea agrees we could also upload the test db to the ftp and I could make a test with that.
> 
> Think I've addressed all of your comments Vangelis. Just seen that the nf-test has failed - made a new output file that it's never made before even though I've run it countless times 🤦‍♀️ will sort that next week.

You can also just add it here... which I think you already did

mberacochea commented:
@tgurbich can you please take a look at this one please?

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/main.nf:
```diff
@@ -0,0 +1,54 @@
+        --hmm_cpu ${task.cpus} \\
+        --tf_cpu ${task.cpus} \\
+        --dbcan_thread ${task.cpus} \\
+        --db_dir \$(pwd)/dbcan_db \\
```
```suggestion
        --db_dir dbcan_db \\
```
The pwd is not needed here

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/main.nf:
```diff
@@ -0,0 +1,54 @@
+        --dbcan_thread ${task.cpus} \\
+        --db_dir \$(pwd)/dbcan_db \\
+        --out_dir dbcan \\
+        --cgc_substrate \\
```
```suggestion
```

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/tests/main.nf.test:
```diff
@@ -0,0 +1,105 @@
+                "tf-2.out"
+            ]
+
+            file(process.out.dbcan_output.get(0).get(1)).eachFileRecurse { file ->
```
In a real nf-core module, all of these would still go in the snapshot. .exists()  .contains() .match() combinations
I'll let @mberacochea decide if this is good enough for our nf-modules.

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/tests/main.nf.test:
```diff
@@ -0,0 +1,105 @@
+
+        then {
+            assertAll(
+                { assert process.success },
```
```suggestion
                { assert process.success },
                { assert snapshot(process.out).match() }
```
Probably removed by mistake

vagkaratzas commented on modules/ebi-metagenomics/dbcan/dbcan/main.nf:
```diff
@@ -0,0 +1,54 @@
+        --cgc_substrate \\
+        --cluster ${gff} \\
+        ${faa} \\
+        protein
```
If this is only ever going to be run with the `protein` mode, leave as it is.
Else, also (just as `--cgc_substrate`) provide this `protein` flag through ext.args (nextflow.config in the tests folder here, and modules.config in a pipeline)

tgurbich replied:
`--cluster` is also an optional parameter

jmattock5 replied:
I think you need `--cluster` to be able to run `protein` so they'd have to be linked in the config.

Could we just rename this module to dbcanprot then copy&slightly modify the module if in the future we want to run it on one of the other settings?

tgurbich replied:
I believe the `--cluster` flag is only required if you want dbcan to annotate clusters from proteins. It is also able to run on proteins without identifying clusters and then the cluster flag is not needed.

vagkaratzas replied:
If the input is a file, that parameter still needs to be there in the module, and in non protein cases, you will have to input empty `[]` and then build the argument within the module (def before the script).

jmattock5 replied:
You're right Tania 😄 

Guess we need to decide if we'd ever **not** want to predict clusters and substrates.

vagkaratzas reviewed (changes requested):
Final Round!

tgurbich commented on modules/ebi-metagenomics/dbcan/dbcan/meta.yml:
```diff
@@ -0,0 +1,183 @@
+        type: file
+        description: A FASTA file containing query sequences, they can be proteins of
+          contigs
+        pattern: "*.{fasta,faa,fa}"
```
```suggestion
        pattern: "*.{fasta,faa,fa,fna}"
```

vagkaratzas reviewed (approved):
epic

mberacochea commented:
> Just tested with my test workflow and it's all good 🥳
> 
> I've checked and running with an assembly, using the mode prok and args `-c cluster` instead of `--cluster ${gff}` completes successfully. We need to make sure we include this in the README when it's created

We need another test for this.. in the future
